### PR TITLE
Add unit testing to robot hardware

### PR DESCRIPTION
**Pull request by bmagyar**
As per mentioned during the last meeting that this package has no proper unit testing, I've added some for `RobotHardware`. The initial plan was to move the one from `test_robot_hardware` but it is very tied to what exactly is happening with that implementation and I wanted to have a more general set of tests.

There are some changes in this PR that should be credited to uncrustify...

On the other hand, @Karsten1987 if you could assist with the cppcheck failure I'd be glad. I am getting the same locally and don't know what to do with it really...
```
test/test_robot_hardware_interface.cpp:90]: (error: syntaxError) syntax error
```

### CODE REVIEW

**Comment by jordan-palacios** [comment]
> On the other hand, @Karsten1987 if you could assist with the cppcheck failure I'd be glad. I am getting the same locally and don't know what to do with it really...
> 
> ```
> test/test_robot_hardware_interface.cpp:90]: (error: syntaxError) syntax error
> ```

@bmagyar This is the one were it complains on the first usage of `TEST_F` right? I managed to fix it by not having the`TEST_F` inside a namespace.

File: hardware_interface/test/test_robot_hardware_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    return hw::HW_RET_OK;
+  }
+
+  hw::hardware_interface_ret_t write() override
+  {
+    return hw::HW_RET_OK;
+  }
+};
+
+const auto JOINT_NAME = "joint_1";
```
Comment (jordan-palacios) [inline hardware_interface/test/test_robot_hardware_interface.cpp:46]:
```suggestion
constexpr auto JOINT_NAME = "joint_1";
```

File: hardware_interface/test/test_robot_hardware_interface.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+const auto NEW_JOINT_NAME = "joint_2";
```
Comment (jordan-palacios) [inline hardware_interface/test/test_robot_hardware_interface.cpp:47]:
```suggestion
constexpr auto NEW_JOINT_NAME = "joint_2";
```

File: hardware_interface/test/test_robot_hardware_interface.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  void SetUpNewHandles()
+  {
+    robot.register_joint_command_handle(&new_pos_command_handle);
+    robot.register_joint_state_handle(&new_state_handle);
+    robot.register_operation_mode_handle(&new_op_mode_handle);
+  }
+
+  MyTestRobotHardware robot;
```
Comment (jordan-palacios) [inline hardware_interface/test/test_robot_hardware_interface.cpp:74]: I prefer member variables to end with an underscore.

File: hardware_interface/test/test_robot_hardware_interface.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  EXPECT_EQ(hw::HW_RET_ERROR, robot.get_operation_mode_handle(NEW_JOINT_NAME, &op_mode_handle));
+
+  SetUpNewHandles();
+  state_handle = nullptr;
+  EXPECT_EQ(hw::HW_RET_OK, robot.get_joint_state_handle(NEW_JOINT_NAME, &state_handle));
+  cmd_handle = nullptr;
+  EXPECT_EQ(hw::HW_RET_OK, robot.get_joint_command_handle(NEW_JOINT_NAME, &cmd_handle));
+  op_mode_handle = nullptr;
+  EXPECT_EQ(hw::HW_RET_OK, robot.get_operation_mode_handle(NEW_JOINT_NAME, &op_mode_handle));
+}
```
Comment (jordan-palacios) [inline hardware_interface/test/test_robot_hardware_interface.cpp:178]: Maybe also check handles are not null after successful `get_*_handle(...)`.

File: hardware_interface/src/joint_state_handle.cpp [hunk 5]
```diff
@@ -21,18 +21,13 @@
 
 namespace hardware_interface
 {
-
 JointStateHandle::JointStateHandle()
-: name_(),
-  pos_(nullptr),
-  vel_(nullptr),
-  eff_(nullptr)
-{}
```
Comment (Karsten1987) [inline hardware_interface/src/joint_state_handle.cpp:30]: why these changes? They seem unrelated

File: hardware_interface/src/joint_command_handle.cpp [hunk 6]
```diff
@@ -21,38 +21,39 @@
 
 namespace hardware_interface
 {
-
 JointCommandHandle::JointCommandHandle()
-: name_(),
-  cmd_(nullptr)
-{}
```
Comment (Karsten1987) [inline hardware_interface/src/joint_command_handle.cpp:28]: why these changes? They seem unrelated

File: hardware_interface/src/joint_command_handle.cpp [hunk 7]
```diff
@@ -21,38 +21,39 @@
...
+: name_(), cmd_(nullptr)
+{
+}
 
-JointCommandHandle::JointCommandHandle(
-  const std::string & name,
-  double * cmd)
```
Comment (Karsten1987) [inline hardware_interface/src/joint_command_handle.cpp:32]: why these changes? They seem unrelated

File: hardware_interface/src/joint_command_handle.cpp [hunk 8]
```diff
@@ -21,38 +21,39 @@
...
+const std::string & JointCommandHandle::get_name() const
 {
   return name_;
 }
 
-double
-JointCommandHandle::get_cmd() const
+double JointCommandHandle::get_cmd() const
 {
+  THROW_ON_NULLPTR(cmd_)
```
Comment (Karsten1987) [inline hardware_interface/src/joint_command_handle.cpp:42]: I disagree on this one. We should check the pointer before de-referencing it

File: hardware_interface/src/joint_state_handle.cpp [hunk 9]
```diff
@@ -21,18 +21,13 @@
...
+: name_(), pos_(nullptr), vel_(nullptr), eff_(nullptr)
+{
+}
 
 JointStateHandle::JointStateHandle(
-  const std::string & name,
-  const double * pos,
-  const double * vel,
+  const std::string & name, const double * pos, const double * vel,
```
Comment (Karsten1987) [inline hardware_interface/src/joint_state_handle.cpp:30]: either line break after all, or no line break at all. This one looks weird.

File: hardware_interface/src/joint_state_handle.cpp (hunk 5 above)
Comment (bmagyar) [reply hardware_interface/src/joint_state_handle.cpp:30]: I really don't know, as I put in the opening post, 
> There are some changes in this PR that should be credited to uncrustify...

Sure, I'll revert 'em

File: hardware_interface/test/test_robot_hardware_interface.cpp (hunk 1 above)
Comment (bmagyar) [reply hardware_interface/test/test_robot_hardware_interface.cpp:46]: done

File: hardware_interface/test/test_robot_hardware_interface.cpp (hunk 3 above)
Comment (bmagyar) [reply hardware_interface/test/test_robot_hardware_interface.cpp:74]: underscored all of 'em

File: hardware_interface/test/test_robot_hardware_interface.cpp (hunk 4 above)
Comment (bmagyar) [reply hardware_interface/test/test_robot_hardware_interface.cpp:178]: good idea! added a new test for this

File: hardware_interface/src/joint_command_handle.cpp (hunk 8 above)
Comment (bmagyar) [reply hardware_interface/src/joint_command_handle.cpp:42]: Could you please elaborate on this one? This is checking the pointer before de-referencing it and throwing if null. These checks are required because there is a way to create handles with nullptr inside

File: hardware_interface/src/joint_state_handle.cpp (hunk 9 above)
Comment (bmagyar) [reply hardware_interface/src/joint_state_handle.cpp:30]: all uncrustify, I'll revert the unnecessary style changes

**Comment by bmagyar** [comment]
@Karsten1987 all unsolicited style changes have been reverted but I have to admit I kinda like them better that way, may submit a PR later with just that. One outstanding point from you regarding the de-referencing comment. 

@jordan-palacios thanks for the note on `TEST_F`, that fixed it!

File: hardware_interface/CMakeLists.txt [hunk 10]
```diff
@@ -52,9 +52,15 @@ if(BUILD_TESTING)
...
-  ament_add_gtest(test_macros test/test_macros.cpp)
+  find_package(ament_cmake_gmock REQUIRED)
+  ament_add_gmock(test_macros test/test_macros.cpp)
   target_include_directories(test_macros PRIVATE include)
   ament_target_dependencies(test_macros rcpputils)
+
+  ament_add_gmock(test_robot_hardware_interfaces test/test_robot_hardware_interface.cpp)
+  target_include_directories(test_robot_hardware_interfaces PRIVATE include)
+  target_link_libraries(test_robot_hardware_interfaces hardware_interface)
+
```
Comment (Karsten1987) [inline hardware_interface/CMakeLists.txt:63]:
```suggestion
```

File: hardware_interface/include/hardware_interface/joint_command_handle.hpp [hunk 11]
```diff
@@ -55,6 +55,9 @@ class JointCommandHandle
   void
   set_cmd(double cmd);
 
+  HARDWARE_INTERFACE_PUBLIC
+  bool valid_pointers() const;
```
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/joint_command_handle.hpp:59]: that function name should be more precise. `valid_pointers` doesn't really say anything about what that function does, meaning which pointers are being checked.

File: hardware_interface/src/joint_command_handle.cpp (hunk 8 above)
Comment (Karsten1987) [reply hardware_interface/src/joint_command_handle.cpp:47]: Must have been a visual thing. The change here looks good to me.

File: hardware_interface/include/hardware_interface/joint_command_handle.hpp (hunk 11 above)
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/joint_command_handle.hpp:59]: also, the function states `pointers` as in multiple pointers, the implementation does however only check a single `cmd_`.

Comment (bmagyar) [reply hardware_interface/include/hardware_interface/joint_command_handle.hpp:59]: I have the same function implemented in all the other handles and I call this from the template function so they should all be called the same. TBH a proper solution would see inheriting such an interface to really ensure this function existing the same way for all
